MudExpansionPanel: Add toggle keyboard hotkeys, Add ability to remove content from DOM

nccadman19 opened the pull request:
fix(MudExpansionPanel): improve accessibility for keyboard and mouse interaction

- Makes header focusable and responsive to Enter/Space keys
- Prevents tab navigation into collapsed panel content
- Aligns behavior with accessibility standards (WCAG)

Tested locally in MudBlazor.Docs to confirm behaviour across input methods.

## Description

This PR enhances accessibility in `MudExpansionPanel` when using custom `TitleContent`. The default behavior does not support keyboard interaction or proper focus handling, particularly when the panel is collapsed. This results in users being unable to toggle the panel via keyboard or inadvertently tabbing into visually hidden content.

**Changes include:**
- Making the header focusable when custom content is used
- Allowing toggling via `Enter` and `Space`
- Preventing tab order from entering content when collapsed
- Preserving click behavior on both container and text elements

These changes improve usability for keyboard and screen reader users and follow accessibility best practices.

## How Has This Been Tested?

Tested locally by running the MudBlazor docs site using the edited source code:
- Verified focus behavior and keyboard toggling with `Tab`, `Enter`, and `Space`
- Confirmed tab order skips collapsed content
- Manually tested pointer interactions with both whitespace and inner text
- No console errors and no impact to existing styles or features

## Type of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

## Checklist

- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests (manual/visual verification in Docs).

## Related Issues
- Closes #11530

nccadman19 committed:
BugFix: Mud Expansion Panel

fix(MudExpansionPanel): improve accessibility for keyboard and mouse interaction

- Makes header focusable and responsive to Enter/Space keys
- Prevents tab navigation into collapsed panel content
- Aligns behavior with accessibility standards (WCAG)

Tested locally in MudBlazor.Docs to confirm behaviour across input methods.

ScarletKuro commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor:
```diff
@@ -19,8 +19,11 @@
-        <div class="@PanelContentClassname">
-            @ChildContent
-        </div>
+		@if (_expandedState.Value)
```
@danielchalmers do we want to hide the content from the DOM?

danielchalmers replied:
This would cause it to re-render excessively and would be a breaking change.

nccadman19 replied:
Hey, sorry never done this before! Pushed a new change, would that work??

```
@if (!RemoveContentOnCollapse || _expandedState.Value)
        {
            <div class="@PanelContentClassname" hidden="@(!_expandedState.Value)" aria-hidden="@(!_expandedState.Value)">
                @ChildContent
            </div>
        }

 [Parameter]
        [Category(CategoryTypes.ExpansionPanel.Behavior)]
        public bool RemoveContentOnCollapse { get; set; } = false;

private async Task HandleKeyDown(KeyboardEventArgs e)
        {
            if(Disabled)
            {
                return;
            }
            
            if (e.Key == "Enter" || e.Key == " ")
            {
                await ToggleExpansionAsync();
            }
        }
```

danielchalmers commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor:
```diff
@@ -2,7 +2,7 @@
 
 <div @attributes="UserAttributes" class="@Classname" style="@Style">
-    <div class="@HeaderClassname" @onclick="ToggleExpansionAsync">
+	<div class="@HeaderClassname" @onclick="ToggleExpansionAsync" tabindex="0" @onkeydown="HandleKeyDown">
```
It's possible to tab to disabled panels

danielchalmers replied:
Also the code's using tabs instead of spaces

nccadman19 replied:
Updated new PR to spaces and added this to fix the disabled tabbing issue

```
<div class="@HeaderClassname" @onclick="ToggleExpansionAsync" tabindex="@(Disabled ? null : "0")" @onkeydown="HandleKeyDown">
```

danielchalmers reviewed (changes requested):
Keyboard accessibility is always needed! Just some things to tweak

nccadman19 committed:
feat(MudExpansionPanel): prevent tab access to disabled panels and add optional content removal

- Prevented focus and keyboard interaction on disabled panel headers by conditionally omitting tabindex
- Introduced `RemoveContentOnCollapse` parameter to optionally remove panel content from the DOM when collapsed
- Default behaviour retains content to preserve component state and avoid breaking changes
- Added `hidden` and `aria-hidden` attributes to hide content from accessibility tree while keeping it mounted

nccadman19 committed:
Formatting error fix

nccadman19 committed:
Merge branch 'dev' into fix/accessible-expansion-toggle

danielchalmers commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor.cs:
```diff
@@ -154,6 +155,16 @@ public partial class MudExpansionPanel : MudComponentBase, IDisposable
+        /// </remarks>
+        [Parameter]
+        [Category(CategoryTypes.ExpansionPanel.Behavior)]
+        public bool RemoveContentOnCollapse { get; set; } = false;
```
Let's flip the bool and call it `KeepContentAlive` (true by default) similar to MudTabs

ScarletKuro commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor:
```diff
@@ -19,8 +19,11 @@
-        <div class="@PanelContentClassname">
-            @ChildContent
-        </div>
+        @if (!RemoveContentOnCollapse || _expandedState.Value)
```
@danielchalmers is that logic correct? shouldn't this be `&`?

danielchalmers replied:
If `!RemoveContentOnCollapse` (default) then it's always rendered. Otherwise only when expanded

danielchalmers replied:
But it does bring up a good reasons we should have tests to cover this

danielchalmers commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor:
```diff
@@ -19,8 +19,11 @@
-        </div>
+        @if (!RemoveContentOnCollapse || _expandedState.Value)
+        {
+            <div class="@PanelContentClassname" hidden="@(!_expandedState.Value)" aria-hidden="@(!_expandedState.Value)">
```
I don't believe you should use both `hidden` and `aria-hidden` at the same time

danielchalmers commented on src/MudBlazor/Components/ExpansionPanel/MudExpansionPanel.razor.cs:
```diff
@@ -253,5 +264,18 @@ protected virtual void Dispose(bool disposing)
             }
         }
+
+        private async Task HandleKeyDown(KeyboardEventArgs e)
```
Should be renamed HandleKeyDown -> HandleKeyDownAsync

nccadman19 committed:
Property and method named changed as requested
Aria hidden removed
Flipped bool similar to MudTabs

danielchalmers reviewed (changes requested):
Awesome! Could you add some HandleKeyDownAsync and KeepContentAlive tests to finish it off? (also small thing but its back to tabs in the razor)

nccadman19 committed:
test: add comprehensive tests for MudExpansionPanel keyboard and accessibility features

- Add keyboard navigation tests for Enter/Space key toggling
- Test accessibility behavior with hidden attribute for collapsed content
- Add HandleKeyDownAsync tests for disabled state and non-trigger keys
- Add KeepContentAlive tests for content rendering lifecycle
- Verify proper interaction between KeepContentAlive and accessibility states

MudExpansionPanel updated to spaces

nccadman19 commented:
> Awesome! Could you add some HandleKeyDownAsync and KeepContentAlive tests to finish it off? (also small thing but its back to tabs in the razor)

Hey, tests added! Hope they're ok, and updated the files back to spaces

danielchalmers reviewed (approved):
Great work!
